jax/interpreters/pxla.py

```diff
@@ -681,32 +681,45 @@ def dynamic_fun(dummy, *args):
     # XLA computation at all; we handle this as a special case so we can stage
     # out multi-replica XLA computations regardless of the hardware available.
     # The 'None' values here are just dummies we know will be ignored.
-    handlers = [_pval_to_result_handler(axis_size, None, pval, local_devices,
-                                        backend)
-                for pval in out_pvals]
+    handlers = [
+        _pval_to_result_handler(axis_size, None, None, None, pval, local_devices,
+                                backend) for pval in out_pvals
+    ]
     results = [handler(None) for handler in handlers]
     return lambda *_: results
 
   jaxpr_replicas = xla.jaxpr_replicas(jaxpr)
   num_local_replicas = axis_size * jaxpr_replicas
   num_global_replicas = global_axis_size * jaxpr_replicas
+  arg_parts, out_parts, num_partitions = _find_partitions(jaxpr)
+
+  num_local_shards = num_local_replicas * num_partitions
+  num_global_shards = num_global_replicas * num_partitions
+
   axis_env = xla.AxisEnv(num_global_replicas, (axis_name,), (global_axis_size,), devices)
 
   tuple_args = len(sharded_avals) > 100  # pass long arg lists as tuple for TPU
 
   c = xb.make_computation_builder("pmap_{}".format(fun.__name__))
   xla_consts = _map(partial(xb.constant, c), consts)
   replicated = [not m for m in mapped_invars]
-  xla_args = xla._xla_callable_args(c, sharded_avals, tuple_args, replicated)
+  xla_args = xla._xla_callable_args(c, sharded_avals, tuple_args, replicated,
+                                    arg_parts)
   out_nodes = xla.jaxpr_subcomp(c, jaxpr, backend, axis_env, xla_consts,
                                 extend_name_stack(wrap_name(name, 'pmap')), *xla_args)
-  built = c.build(xops.Tuple(c, out_nodes))
+  build_out_tuple = partial(xops.Tuple, c, out_nodes)
+  if out_parts is not None:
+    out_tuple = xb.with_sharding(c, out_parts, build_out_tuple)
```

Comment:
Wouldn't this be nicer if `xb.with_sharding` were a context manager, and it could take `None` with the SpatialSharding?

Reply:
Context manager seems like a good idea, and I might also push the to-shard-or-not-to-shard logic into it, since it's pretty awkward. I'm gonna do this as a follow-up PR, I filed https://github.com/google/jax/issues/3183 for now.